### PR TITLE
message retry persistance and aphrodite support

### PR DESCRIPTION
**Pull request by ewof**
should note that for aphrodite, when u regen a message and u swipesPerGeneration > 1 it replaces all the retries with the new gens (intended)

### CODE REVIEW

File: web/shared/GenerationSettings.tsx [hunk 1]
```diff
@@ -877,10 +901,36 @@ const GenSettings: Component<Props & { pane: boolean; format?: ThirdPartyFormat;
...
           aiSetting={'repetitionPenaltySlope'}
           format={props.format}
         />
+        <RangeInput
+          fieldName="etaCutoff"
+          label="ETA Cutoff"
+          helperText="In units of 1e-4; a reasonable value is 3. The main parameter of the special Eta Sampling technique. See [this paper](https://arxiv.org/pdf/2210.15191.pdf) for a description."
+          min={0}
+          max={20}
+          step={0.01}
```
Comment (AlpinDale) [inline web/shared/GenerationSettings.tsx]: The steps should be 1e-4 or `0.0001`. I'd keep the max at 5, honestly. For both of them.

File: web/shared/GenerationSettings.tsx [hunk 2]
```diff
@@ -877,10 +901,36 @@ const GenSettings: Component<Props & { pane: boolean; format?: ThirdPartyFormat;
...
+          aiSetting={'etaCutoff'}
+          format={props.format}
+        />
+        <RangeInput
+          fieldName="epsilonCutoff"
+          label="Epsilon Cutoff"
+          helperText="In units of 1e-4; a reasonable value is 3. This sets a probability floor below which tokens are excluded from being sampled."
+          min={0}
+          max={9}
+          step={0.01}
```
Comment (AlpinDale) [inline web/shared/GenerationSettings.tsx]: Same here.

**Comment by AlpinDale** [comment]
It looks like you need to manually enter the model name. You can grab it by sending a GET request to `/v1/models` route. A typical response would look like this:

```json
{"object":"list","data":[{"id":"goliath","object":"model","created":1701961921,"owned_by":"pygmalionai","root":"goliath","parent":null,"permission":[{"id":"modelperm-46a6d814213b45fbb371d60695cf5fff","object":"model_permission","created":1701961921,"allow_create_engine":false,"allow_sampling":true,"allow_logprobs":true,"allow_search_indices":false,"allow_view":true,"allow_fine_tuning":false,"organization":"*","group":null,"is_blocking":false}]}]}
```

The name is `goliath` in this example. The GET request would need to include the API key in its header.

**Comment by AlpinDale** [comment]
I'm still getting that bug where multi-gen combines both generations into one message for me (looks all mangled), but @ewof has been unable to reproduce it. Otherwise, this PR LGTM.

**Comment by AlpinDale** [comment]
I've narrowed down the issue to streaming responses only. It seems to interleave responses if you enable streaming. A possible solution is to only stream index 0 response and accumulate the others in the background, then finally allow swiping once they've all finished streaming. For reference, this is what the JSON response looks like for two requested outputs:
```
data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": " but", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": " but", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": " it", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": " you", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": " would", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": " short", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": " seem", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": " lived", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": " that", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": " ones", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": " these", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 0, "text": "", "logprobs": null, "finish_reason": "length"}], "usage": {"prompt_tokens": 25, "total_tokens": 31, "completion_tokens": 6}}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": " seem", "logprobs": null, "finish_reason": null}]}

data: {"id": "cmpl-db8cd384056845ffa317c723552c537b", "created": 1702428590, "model": "goliath", "choices": [{"index": 1, "text": "", "logprobs": null, "finish_reason": "length"}], "usage": {"prompt_tokens": 25, "total_tokens": 31, "completion_tokens": 6}}

data: [DONE]
```

The `[DONE]` response would be the indicator that the request has finished generating.

**Comment by ewof** [comment]
pushed fix for this

**Comment by AlpinDale** [comment]
That seems to have fixed it but there's a new issue: it inserts the string `undefined` at the start of each message - applicable to all swipes.
image
image

**Review by AlpinDale — APPROVED** [review]
LGTM.

**Comment by sceuick** [comment]
Will be getting through the backlog of PRs over the next few days. Thanks @ewof and @AlpinDale :+1:

**Pull request merged**